halo2_poseidon/src/lib.rs

```diff
@@ -176,16 +188,82 @@ impl<F, const RATE: usize> SpongeMode for Squeezing<F, RATE> {}
 
 impl<F: fmt::Debug, const RATE: usize> Absorbing<F, RATE> {
     pub(crate) fn init_with(val: F) -> Self {
+        let mut state = Self::init_empty();
+        state.absorb(val).expect("state is not full");
+        state
+    }
+
+    /// Initializes an empty sponge in the absorbing state.
+    pub fn init_empty() -> Self {
         Self(
-            iter::once(Some(val))
-                .chain((1..RATE).map(|_| None))
+            (0..RATE)
+                .map(|_| None)
                 .collect::<Vec<_>>()
                 .try_into()
                 .unwrap(),
         )
     }
 }
 
+impl<F, const RATE: usize> Absorbing<F, RATE> {
+    /// Attempts to absorb a value into the sponge state.
+    ///
+    /// Returns the value if it was not absorbed because the sponge is full.
+    pub fn absorb(&mut self, value: F) -> Result<(), F> {
```

Comment:
Is `Result` the right type to return here? Or should it be a bespoke enum? It looks to me like the kind of thing that I would ordinarily use `Either` for rather than `Result`, but a bespoke enum might be better.